### PR TITLE
Support zero-width integer literals.

### PR DESCRIPTION
**Pull request by dtzSiFive**
These are in the FIRRTL language specification now and supported by both SFC + MFC.

If integer literal width is specified, support zero-width (`0.U(0.W)`), otherwise at least 1-bit (`0.U`), for compatibility with existing code.

Modified `assertKnownWidth`/`assertInferredWidth` to allow checking for width of zero, avoid attempting to use `-1S(0.W)`.  Used these to check basic zero-width behavior of wires/regs being initialized with zero-width integer literals.

Add a few tests to ensure zero-width literals and a few operations/constructs work as expected, and that their behavior is preserved.

### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [ ] Did you state the API impact?
- [ ] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- new feature/API

#### Desired Merge Strategy

- Squash: The PR will be squashed and merged

#### Release Notes

Add support for zero-width integer literals, which must be explicitly requested (`0.U(0.W)`), for compatibility reasons `0.U` will still be (minimum) 1-bit.

### Reviewer Checklist (only modified by reviewer)
- [x] Did you add the appropriate labels?
- [x] Did you mark the proper milestone (Bug fix: `3.4.x`, [small] API extension: `3.5.x`, API modification or big change: `3.6.0`)?
- [x] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

**Comment by dtzSiFive** [comment]
At least some of the CI failures are because current firtool used by CI is not new enough, PR bumping to latest submitted: #2933 .

Looks like really want `0.U` to be 1-bit, will look into making that the case next!

File: src/test/scala/chiselTests/WidthSpec.scala [hunk 1]
```diff
@@ -67,6 +97,24 @@ abstract class WireRegWidthSpecImpl extends ChiselFlatSpec {
...
+      val w = builder(new ZeroWidthBundle)
+      w := ZeroWidthBundle.intoWire()
+      w.y
+    }
+    assertInferredWidth(1) {
+      val w = builder(Vec(1, UInt()))
+      w(0) := 0.U(0.W)
+      w(0)
+    }
+  }
```
Comment (jackkoenig) [inline src/test/scala/chiselTests/WidthSpec.scala:117]: This seems wrong to me (and super kudos for writing these tests because I would not have noticed this behavior without them).

I would expect `Cat(1.U, 0.U(0.W), 1.U)` to be isomorphic to `Cat(1.U, WireInit(UInt(), 0.U(0.W)), 1.U)`, but this suggests  otherwise, why?
This indeed seems to be the case when I run that actual Chisel to Verilog with both MFC and SFC so I'm confused why there's inferred width == 1?

Is it perhaps a test setup issue? Determining inferred width from Chisel is tricky and maybe it has a bug around zero-width.

**Review by jackkoenig — COMMENTED** [review]
This looks really good and I especially appreciate all of the tests. There is 1 test that isn't making sense to me but otherwise this LGTM.

File: src/test/scala/chiselTests/WidthSpec.scala (hunk 1 above)
Comment (dtzSiFive) [reply src/test/scala/chiselTests/WidthSpec.scala:117]: > This seems wrong to me

Me too, honestly, thanks for bringing it up.  I agree with your suggestion it may be a test setup issue (definitely was not 100% on those changes when I made them) and indeed it seems to have been!  Will push a fix soon, happily it was actually really dumb with fresh eyes -- was using a `0.U` which this PR went out of the way to make NOT zero-width (min 1-bit) which was why it wouldn't pass things that actually were zero-width.  Making it `0.U(0.W)` fixes that :+1: .

> I would expect `Cat(1.U, 0.U(0.W), 1.U)` to be isomorphic to `Cat(1.U, WireInit(UInt(), 0.U(0.W)), 1.U)`, but this suggests otherwise, why?

Fantastic test case, thanks!

Is this something suitable for a test as well? Not sure where / how that might work, especially without assuming constant propagation in the compiler and such, but for I really liked that (immediately seemed clear what you were trying to check/determine/say) and so would be great to check moving forward.  If that's not really suitable, the width checks already present are probably enough.  Who tests the testers, indeed.

( I've been testing various behavior using an increasingly modified blink example, so I can replicate your findings re:`2'h3` for both (no surprise, for my own sanity) )

Comment (jackkoenig) [reply src/test/scala/chiselTests/WidthSpec.scala]: > Is this something suitable for a test as well? Not sure where / how that might work, especially without assuming constant propagation in the compiler and such, but for I really liked that (immediately seemed clear what you were trying to check/determine/say) and so would be great to check moving forward. If that's not really suitable, the width checks already present are probably enough.

Yeah I have the same thoughts/questions/concerns. I think it would be really handy to be able to express things like this, and perhaps a suite of formal checks for certain properties we expect to hold would be good. I think such a thing is beyond the scope of this PR though--I think the tests you've written are more than sufficient. Nice work 🙂

**Review by jackkoenig — APPROVED** [review]
LGTM! Nice work and great tests!
